site/content/blog/2020-09-07-improving-performance-through-load-testing.md

```diff
@@ -0,0 +1,83 @@
+
+Fortunately, here at Mattermost we have a convenient advantage: we use our own application in our day to day work. This means we are able to continuously collect a considerable amount of realistic metrics from our community servers. We can then use such data to model user behaviour. To do this we create a table of user's actions with their respective frequencies. This action-frequency map will feed the simulation process: a simulated user will randomly pick and perform an action based on its frequency so that most common actions will be picked more often.
+
+![actions-map](/blog/2020-09-07-improving-performance-through-load-testing/actions-map.png)
+
+In such an example, the chance of a user switching a channel will be double the chance of creating a new post and four times that of replying to a thread.
+
+### Getting some answers
+
+Now that we have a simulation in place, we come back to the original question of this blog post: how can we make our load testing software tell us how many concurrently active users our target application can support?
+
+Instead of relying on [magic formulas](https://en.wikipedia.org/wiki/Magic_number_(programming)) which are hard to explain (and sometime defend) or on endless and error-prone manual testing, we thought of picking an idea out of [control theory](https://en.wikipedia.org/wiki/Control_theory) that could give us both automation and precision at the same time: it's called a **feedback loop**.
+
+![feedback-loop](/blog/2020-09-07-improving-performance-through-load-testing/feedback-loop.png)
+
+The idea is quite simple: have the load testing engine react to what's happening on the target system by dynamically changing the amount of load applied in response to some event. This is achieved through the monitoring of those metrics that can best signal service degradation. Some that we make use of are:
+
+- Request durations (average and percentiles)
+- Request errors rate
+- Client timeouts rate
+
+![graph](/blog/2020-09-07-improving-performance-through-load-testing/graph.png)
+
+The load testing software constantly increases the number of active users up until it detects signs of performance degradation. This happens when the metrics being monitored reach their configured thresholds (e.g. 100ms for average request duration).
+In response to such an event, the load testing engine starts to lower the number of users. This process continues for a while until an **equilibrium point** is reached. This value, over which performance starts to degrade, gives us a fairly good estimate on the users capacity of our target application.
+
+## Wrapping Up
+
+The success of a software system greatly depends on its performance. In today's cloud-oriented world this is increasily more true. A small outage can have a serious impact on your customers and eventually affect your company's revenue, especially when a [service level agreement](https://en.wikipedia.org/wiki/Service_level_agreement) is in place.
```

Comment:
```suggestion
The success of a software system greatly depends on its performance, especially in today's cloud-oriented world. A small outage can have a serious impact on your customers and eventually affect your company's revenue, especially when a [service level agreement](https://en.wikipedia.org/wiki/Service_level_agreement) is in place.
```

Reply:
@justinegeffen 

Any suggestion to avoid the consecutive use of "especially" in the paragraph?

Reply:
The second especially can be "above all" 🤷